agent/config/file_watcher.go

```diff
@@ -0,0 +1,225 @@
+	}
+}
+
+func (w *FileWatcher) Add(filename string) error {
+	timeout := time.After(timeoutDuration)
+
+	// explicitly do not support symlink as the behaviour is not consistent between OSs
+	if isSymLink(filename) {
+		return fmt.Errorf("symbolic link are not supported %s", filename)
+	}
+	select {
+	case w.toBeAdded <- filename:
+		return nil
+	case <-timeout:
+		return fmt.Errorf("file add timedout %s", filename)
+	}
+}
+
+func isSymLink(filename string) bool {
+	symlinks, err := os.Readlink(filename)
+	if err == nil && symlinks != filename {
+		return true
+	}
+	return false
+}
+
+func (w *FileWatcher) Close() error {
+	close(w.done)
+	return w.watcher.Close()
+}
+
+func (w *FileWatcher) watch() {
+	timer := time.NewTimer(w.reconcileTimeout)
+	defer timer.Stop()
+	for {
+		select {
+		case event, ok := <-w.watcher.Events:
+			if !ok {
+				w.logger.Error("watcher event channel is closed")
+				return
+			}
+			w.logger.Debug("received watcher event", "event", event)
+			if err := w.handleEvent(event); err != nil {
+				w.logger.Error("error handling watcher event", "error", err, "event", event)
+			}
+			timer.Reset(w.reconcileTimeout)
+		case _, ok := <-w.watcher.Errors:
+			if !ok {
+				w.logger.Error("watcher error channel is closed")
+				return
+			}
+			timer.Reset(w.reconcileTimeout)
+		case <-timer.C:
+			w.reconcile()
+			timer.Reset(w.reconcileTimeout)
+		case filename := <-w.toBeAdded:
+			err := w.add(filename)
+			if err != nil {
+				w.logger.Error("error adding a file", "file", filename, "err", err)
+			}
+		case filename := <-w.toBeRemoved:
+			err := w.remove(filename)
+			if err != nil {
+				w.logger.Error("error removing a file", "file", filename, "err", err)
+			}
+		case <-w.done:
+			return
+		}
+	}
+
+}
+
+func (w *FileWatcher) handleEvent(event fsnotify.Event) error {
+	w.logger.Info("event received ", "filename", event.Name, "OP", event.Op)
+	// we only want Create and Remove events to avoid triggering a relaod on file modification
+	if !isCreate(event) && !isRemove(event) {
+		return nil
+	}
+	configFile, basename, ok := w.isWatched(event.Name)
+	if !ok {
+		return fmt.Errorf("file %s is not watched", event.Name)
+	}
+
+	// we only want to update inode and re-add if the event is on the watched file itself
+	if event.Name == basename {
+		if isRemove(event) {
+			// If the file was removed, set it to be re-added to watch when created
+			err := w.watcher.Add(event.Name)
+			if err != nil {
+				configFile.watched = false
+				configFile.iNode = 0
+				return nil
+			}
+		}
+
+		id, err := w.getFileId(event.Name)
+		if err != nil {
+			return err
+		}
+
+		w.logger.Info("set id ", "filename", event.Name, "id", id)
+		configFile.iNode = id
+		return w.handleFunc(&WatcherEvent{Filename: event.Name})
+	}
+	if isCreate(event) {
+		return w.handleFunc(&WatcherEvent{Filename: event.Name})
+	}
+	return nil
+}
+
+func (w *FileWatcher) isWatched(filename string) (*watchedFile, string, bool) {
```

Comment:
This method will terminate right? 🤔 at some point we will hit a `dir` that has to be watched

Reply:
Oh I fixed something in here but I don't see it, I think I did it in the wrong place!! Let me bring it here.

Reply:
Should be better now